rust/src/quic/quic.rs

```diff
@@ -112,44 +129,146 @@ impl QuicState {
         return None;
     }
 
-    fn parse(&mut self, input: &[u8]) -> bool {
-        match QuicHeader::from_bytes(input, DEFAULT_DCID_LEN) {
-            Ok((rest, header)) => match QuicData::from_bytes(rest) {
-                Ok(data) => {
-                    // no tx for the short header (data) frames
-                    if header.ty != QuicType::Short {
-                        let mut sni : Option<Vec<u8>> = None;
-                        let mut ua : Option<Vec<u8>> = None;
-                        for frame in &data.frames {
-                            if let Frame::Stream(s) = frame {
-                                if let Some(tags) = &s.tags {
-                                    for (tag, value) in tags {
-                                        if tag == &StreamTag::Sni {
-                                            sni = Some(value.to_vec());
-                                        } else if tag == &StreamTag::Uaid {
-                                            ua = Some(value.to_vec());
+    fn decrypt<'a>(
+        &mut self, to_server: bool, header: &QuicHeader, framebuf: &'a [u8], buf: &'a [u8],
+        hlen: usize, output: &'a mut Vec<u8>,
+    ) -> bool {
+        if let Some(keys) = &self.keys {
+            let hkey = if to_server {
+                &keys.remote.header
+            } else {
+                &keys.local.header
+            };
+            if framebuf.len() < 4 + hkey.sample_len() {
+                return false;
+            }
+            let mut h2 = Vec::with_capacity(hlen + header.length);
```

Comment:
what is the max size we might alloc here?

Reply:
Not more than the UDP payload size...
Should it be more obvious or debug tested ?

Reply:
yeah, I was trying to figure out where it is set and the logic is quite hard to follow but ultimately seems to come from `quic_var_uint`, so it might be higher if not properly checked?

If it shouldn't be > UDP payload size, could it be `u16`?

Reply:
It comes indeed from `quic_var_uint`
But then, we have `if header.length > rest.len() {`
before `let (mut framebuf, next_buf) = rest.split_at(header.length);

Do you want the check to happen earlier ?